File: src/main/java/com/github/unchama/seichiassist/data/player/MineStack.kt
package com.github.unchama.seichiassist.data.player

import com.github.unchama.seichiassist.minestack.MineStackObj
import java.util.*

class MineStack(objectCountMap: Map<MineStackObj, Long>) {

  private val objectCountMap = HashMap(objectCountMap)

  constructor() : this(HashMap())

  fun getStackedAmountOf(mineStackObj: MineStackObj): Long = objectCountMap[mineStackObj] ?: 0

  private fun setStackedAmountOf(mineStackObj: MineStackObj, to: Long) {
    objectCountMap[mineStackObj] = to
  }

  /**
   * 指定されたマインスタックオブジェクトのスタック数を増加させる。
   */
  fun addStackedAmountOf(mineStackObj: MineStackObj, by: Long) =
      setStackedAmountOf(mineStackObj, getStackedAmountOf(mineStackObj) + by)

  /**
   * 指定されたマインスタックオブジェクトのスタック数を減少させる。
   */
  fun subtractStackedAmountOf(mineStackObj: MineStackObj, by: Long) =
      addStackedAmountOf(mineStackObj, -by)

}

File: src/main/scala/com/github/unchama/seichiassist/data/player/MineStack.scala
package com.github.unchama.seichiassist.data.player

import com.github.unchama.seichiassist.subsystems.minestack.domain.MineStackObject
import scala.collection.mutable

class MineStack(_objectCountMap: collection.Map[MineStackObject, Long] = collection.Map.empty) {

  private val objectCountMap = mutable.Map.empty ++ _objectCountMap

  /**
   * 指定されたマインスタックオブジェクトのスタック数を減少させる。
   */
  def subtractStackedAmountOf(mineStackObj: MineStackObject, by: Long): Unit =
    addStackedAmountOf(mineStackObj, -by)

  /**
   * 指定されたマインスタックオブジェクトのスタック数を増加させる。
   */
  def addStackedAmountOf(mineStackObj: MineStackObject, by: Long): Unit =
    setStackedAmountOf(mineStackObj, getStackedAmountOf(mineStackObj) + by)

  def getStackedAmountOf(mineStackObj: MineStackObject): Long =
    objectCountMap.getOrElse(mineStackObj, 0)

  private def setStackedAmountOf(mineStackObj: MineStackObject, to: Long): Unit = {
    objectCountMap.put(mineStackObj, to)
  }

  def getObjectCounts: Map[MineStackObject, Long] = {
    objectCountMap.toMap
  }

}
